### ISSUE TITLE
TimeToLive vs message disappearance

### ISSUE THREAD

**Issue by janzakrzewskipl** [Azure/azure-storage-net#645]
Hi,

When and how does a message is deleted when TTL expires?

I add message to queue using Azure Storage Explorer and set it should expire in 7 seconds.
After 7 seconds I don't see it in message list in ASE, which is fine.
One thing is bothering me though, why at the bottom ASE displays "Showing 0 of 1 messages in queue".
At the beginning I thought that this message is still alive, it is just invisible for some time, but this state lasts for few hours now. So I am not sure now, is it deleted or just invisible?

**Comment by erezvani1529** [comment]
Hi @janzakrzewskipl ,

This seems to be an storage explorer issue. Are you able to reproduce it with the storage client?
Time to live is indeed for message expiry and should actually delete the message when its time is up. 

On the other hand, Visibility timeouts will not expire the message and in fact are used for the opposite purpose so that if one processor dies, another processor can pick up the messages.

Thanks!

**Comment by sagoo33** [comment]
I don't believe this is just an ASE issue, as when calling the ApproximateMessageCount property on the client, the expired messages are still shown:

`CloudQueue queueClient = QueueClient.GetQueueReference("test-queue")
qClient.AddMessage("testmessage", new TimeSpan(0,0,2)), null, new QueueRequestOptions() { }, null);
System.Threading.Thread.Sleep(5000);
qClient.FetchAttributes();
var messageCount = qClient.ApproximateMessageCount;`


From what I understand, expired messages should immediately be cleared out of the queue, currently they just look like a message with a visibility timeout...except they never pop back onto the queue.

This currently makes my auto scale rules based on queue message count useless, as it never scales back down due to the expired messages sitting on the queue and never being cleared.


Am I doing something incorrect here?

**Comment by zharenkov** [comment]
Hello, 
Issue is still reproducable.  Message TTL is set to 1 hour. We pushed 2500 messages to the queue. After an hour 1100 of them were not consumed so they should be deleted. But instead of this - they became invisible. Azure Storage Explorer shows "Showing 0 of 1141 messages in queue". My App Service at this moment got 30 running instances and Scale In rule - decrease count to 0 when Approx number of messages in queue will be =0. But this scale in rule won't be triggered, cause messages are still in queue even after TTL. 
![azure_scale_out](https://user-images.githubusercontent.com/5469127/80017380-30d31480-84dd-11ea-928d-cb0434c2d7b0.png)
    
![azure_scale_out_queu](https://user-images.githubusercontent.com/5469127/80017388-34669b80-84dd-11ea-82af-ed9c2be577e6.png)

**Comment by amolb2112** [comment]
Is there any further action planned on this issue ? We plan to use this count for scaling decision. In the absence of this count being reliable, we are compelled to avoid using queue TTL feature.

https://docs.microsoft.com/en-us/answers/questions/299550/expired-messages-not-accessible-but-still-show-in.html?childToView=475981#answer-475981

**Comment by mandrie** [comment]
Any update on this? I am also seeing it, and am wondering if there is an easy way to "clear" the expired messages or the count without impacting the actual messages?
